### DISCUSSION TITLE
Why do Soft Assertions use Proxies underneath?

### DISCUSSION THREAD

**Discussion by tedyoung** [assertj/assertj#3385] [General]
I've been focusing on reducing the runtime of tests and I've noticed that using Soft Assertions slows tests down because of their use of proxying (I know 800ms might not be a long time, but when all of my tests run in under 1500ms, that 800ms is significant). I haven't looked deeply into the code (yet), but I'm wondering what in the design requires the proxies?

**Comment by scordio** [comment]
To get soft assertions, the "hard" assertion implementations are reused and proxied via ByteBuddy. However, this happens at runtime.

We do have #2563 about performance improvements, where we might be able to generate all the proxies at build time via the [`byte-buddy-maven-plugin`](https://github.com/raphw/byte-buddy/tree/master/byte-buddy-maven-plugin), but we haven't had the chance to look at it yet.

Depending on the use case, [`satisfies(Consumer...)`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractAssert.html#satisfies(java.util.function.Consumer...)) might be an alternative to achieve better performance for the execution of multiple assertions.

**Comment by joel-costigliola** [comment]
Yep, code generation would be nice.
